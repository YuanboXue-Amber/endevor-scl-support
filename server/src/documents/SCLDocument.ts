import { TextDocument, Diagnostic, DiagnosticSeverity, TextEdit, CompletionItemKind } from 'vscode-languageserver';
import { ITokenizedString, Tokenizer } from '../parser/Tokenizer';
import { isNullOrUndefined } from 'util';
import { SCLDocumentManager, actionCompletion } from './SCLDocumentManager';
import { match, matchWithoutDiagnose, ParserTags } from '../parser/ParserTags';
import { SETtree, ADDtree, UPDATEtree, DELETEtree, GENERATEtree, MOVEtree, RETRIEVEtree, SIGNINtree, TRANSFERtree,
    APPROVEtree,
    DENYtree,
    BACKINtree,
    BACKOUTtree,
    CASTtree,
    DEFINEPACKAGEtree,
    EXECUTEtree,
    RESETtree,
    COMMITtree,
    LISTtree} from '../parser/syntaxTrees/PrepareTrees';
import { diagnose } from '../parser/syntaxTrees/Parser';
import { IFromTocheck } from '../parser/syntaxTrees/doc/Inode';
import { QUICKFIX_CHOICE_MSG } from '../CodeActionProvider';

/**
 * An interface that extends the vscode diagnostic.
 * starti and endi is the range of diagnostic. This is used when updating document.
 * We first calculate the starti and endi, then update the diagnostic.range
 *
 * @export
 * @interface ISCLDiagnostic
 */
export interface ISCLDiagnostic {
    diagnostic: Diagnostic;
    starti: number;
    endi: number;
}

/**
 * Informaton about one scl statement.
 * Its tokens. Where it start and end. It's diagnostic information
 *
 * @export
 * @class SCLstatement
 */
export class SCLstatement {
    starti: number;
    endi: number;
    tokens: ITokenizedString[] = [];
    diagnostics: ISCLDiagnostic[] = [];
    isRest: boolean = false;

    fromtoCheck: IFromTocheck = {
        from: {
            FILE: false,
            location: {
                ENVIRONMENT: false,
                SYSTEM: false,
                SUBSYSTEM: false,
                TYPE: false,
                STAGE: false,
            }
        },
        to: {
            FILE: false,
            location: {
                ENVIRONMENT: false,
                SYSTEM: false,
                SUBSYSTEM: false,
                TYPE: false,
                STAGE: false,
            }
        },
    };

    /**
     * Creates an instance of SCLstatement from tokens.
     *
     * @param {ITokenizedString[]} tokens
     * @param {number} [startiOffset] The offset for all the indexes in the object to be created
     * @memberof SCLstatement
     */
    constructor(tokens: ITokenizedString[], startiOffset?: number) {
        if (isNullOrUndefined(startiOffset)) {
            startiOffset = 0;
        }

        this.starti = tokens[0].starti + startiOffset;
        this.endi = tokens[tokens.length-1].starti + tokens[tokens.length-1].value.length + startiOffset;
        if (startiOffset === 0) {
            this.tokens = tokens;
        } else {
            tokens.forEach((token) => {
                token.starti += startiOffset as number;
                this.tokens.push(token);
            });
        }
        this.diagnostics = [];
        this.fromtoCheck = {
            from: {
                FILE: false,
                location: {
                    ENVIRONMENT: false,
                    SYSTEM: false,
                    SUBSYSTEM: false,
                    TYPE: false,
                    STAGE: false,
                }
            },
            to: {
                FILE: false,
                location: {
                    ENVIRONMENT: false,
                    SYSTEM: false,
                    SUBSYSTEM: false,
                    TYPE: false,
                    STAGE: false,
                }
            },
        };
        this.isRest = false;
        if (this.tokens.length > 0 && this.tokens[0].value === "REST") {
            this.isRest = true;
            // this.tokens.splice(0, 1);
        }
    }
}

/**
 * Information about an entire document.
 * Its content info. Its scl statements.
 *
 * @export
 * @class SCLDocument
 */
export class SCLDocument {
    textDocument: TextDocument;
    statements: SCLstatement[] = [];
    setCheck: IFromTocheck = {
        from: {
            FILE: false,
            location: {
                ENVIRONMENT: false,
                SYSTEM: false,
                SUBSYSTEM: false,
                TYPE: false,
                STAGE: false,
            }
        },
        to: {
            FILE: false,
            location: {
                ENVIRONMENT: false,
                SYSTEM: false,
                SUBSYSTEM: false,
                TYPE: false,
                STAGE: false,
            }
        },
    };

    constructor(textDocument: TextDocument) {
        this.textDocument = textDocument;
        this.initStatements(); // read the whole content of the document and do synatx diagnose
    }

    /**
     * Tokenize the input string into an array of SCLstatement
     *
     * @private
     * @param {string} text
     * @param {number} [startiOffset] The offset for the current text in the whole document
     * @returns {SCLstatement[]}
     * @memberof SCLDocument
     */
    private parseTextIntoSCLstatementsTokens(text: string, startiOffset?: number): SCLstatement[] {
        const statements: SCLstatement[] = [];
        const tokenizeTextContent: Tokenizer = new Tokenizer(text);
        let token: ITokenizedString = tokenizeTextContent.peekNext();
        let statement: ITokenizedString[] = [];
        while (!isNullOrUndefined(token) && !token.is_eoInput) {
            token = tokenizeTextContent.readNext();

            if (token.is_eoInput) {
                if (statement.length > 0) {
                    statements.push(new SCLstatement(statement, startiOffset));
                }
                break;
            }

            statement.push(token);

            if (token.is_eoStatement) {
                statements.push(new SCLstatement(statement, startiOffset));
                statement = [];
            }
        }
        return statements;
    }

    /**
     * Parse the whole document into an array of scl statements, and do synatx diagnose
     *
     * @private
     * @returns
     * @memberof SCLDocument
     */
    private initStatements() {
        if (isNullOrUndefined(this.textDocument)) {
            this.statements = [];
            return;
        }
        this.statements = this.parseTextIntoSCLstatementsTokens(this.textDocument.getText());
        this.walkStatements();
    }

    /**
     * Walk through all statements in this document to generate all information we need to send to client,
     * e.g. diagnoses, completion, quickfixes, highlights, hovers ...
     *
     * @private
     * @memberof SCLDocument
     */
    private walkStatements() {
        this.statements.forEach((scl) => { this.walkStatement(scl); });
    }

    /**
     * Walk though one statement
     *
     * @private
     * @param {SCLstatement} statement
     * @memberof SCLDocument
     */
    private walkStatement(statement: SCLstatement) {
        let token = statement.tokens[0];
        if (token.value === "REST") {
            if (statement.tokens.length > 1) {
                token = statement.tokens[1];
            } else {
                token.completionItems = [];
                for (const item of actionCompletion) {
                    token.completionItems.push({
                        label: item.toUpperCase() + " ",
                        kind: CompletionItemKind.Function,
                        documentation: "Endevor SCL keyword"
                    });
                }

                this.pushDiagnostic(
                    token, statement,
                    DiagnosticSeverity.Error,
                    `No SCL following label REST`,
                    `${QUICKFIX_CHOICE_MSG}${actionCompletion.join(", ")}`,
                    token);
                return;
            }
        }
        if (match(token, "SET", statement, this))
            diagnose(SETtree, statement, this);
        else if (match(token, "ADD", statement, this))
            diagnose(ADDtree, statement, this);
        else if (match(token, "UPDATE", statement, this))
            diagnose(UPDATEtree, statement, this);
        else if (match(token, "DELETE", statement, this))
            diagnose(DELETEtree, statement, this);
        else if (match(token, "GENERATE", statement, this))
            diagnose(GENERATEtree, statement, this);
        else if (match(token, "MOVE", statement, this))
            diagnose(MOVEtree, statement, this);
        else if (match(token, "RETRIEVE", statement, this))
            diagnose(RETRIEVEtree, statement, this);
        else if (match(token, "SIGNIN", statement, this))
            diagnose(SIGNINtree, statement, this);
        else if (match(token, "TRANSFER", statement, this))
            diagnose(TRANSFERtree, statement, this);

        else if (match(token, "APPROVE", statement, this))
            diagnose(APPROVEtree, statement, this);
        else if (match(token, "DENY", statement, this))
            diagnose(DENYtree, statement, this);
        else if (match(token, "BACKIN", statement, this))
            diagnose(BACKINtree, statement, this);
        else if (match(token, "BACKOUT", statement, this))
            diagnose(BACKOUTtree, statement, this);
        else if (match(token, "CAST", statement, this))
            diagnose(CASTtree, statement, this);
        else if (match(token, "DEFINE", statement, this))
            diagnose(DEFINEPACKAGEtree, statement, this);
        else if (match(token, "EXECUTE", statement, this))
            diagnose(EXECUTEtree, statement, this);
        else if (match(token, "RESET", statement, this))
            diagnose(RESETtree, statement, this);
        else if (match(token, "COMMIT", statement, this))
            diagnose(COMMITtree, statement, this);

        else if (match(token, "LIST", statement, this))
            diagnose(LISTtree, statement, this);

        else {
            this.pushDiagnostic(
                token, statement,
                DiagnosticSeverity.Error,
                `Invalid value specified`,
                `${QUICKFIX_CHOICE_MSG}${actionCompletion.join(", ")}`,
                token);
        }
    }

    fullUpdate() {
        this.setCheck = {
            from: {
                FILE: false,
                location: {
                    ENVIRONMENT: false,
                    SYSTEM: false,
                    SUBSYSTEM: false,
                    TYPE: false,
                    STAGE: false,
                }
            },
            to: {
                FILE: false,
                location: {
                    ENVIRONMENT: false,
                    SYSTEM: false,
                    SUBSYSTEM: false,
                    TYPE: false,
                    STAGE: false,
                }
            },
        };
        if (isNullOrUndefined(this.textDocument)) {
            this.statements = [];
            return;
        }
        this.statements = this.parseTextIntoSCLstatementsTokens(this.textDocument.getText());
        this.walkStatements();
    }

    /**
     * NEVER USED
     *
     * @param {string} newText The new text replacing the original text
     * @param {number} start The start index of the original text
     * @param {number} end The end index of the original text
     * @param {string} origContent The entire original content before change
     * @memberof SCLDocument
     */
    update(newText: string, start: number, end: number, origContent: string) {
        const newTextUp = newText.toUpperCase();
        const origContentUp = origContent.toUpperCase();
        if ((newTextUp.match(/\b(SET)\b/) &&
             (newTextUp.match(/\b(FRO(M|\b))\b/) ||
              newTextUp.match(/\b(TO)\b/)) ) ||
            (origContentUp.match(/(.*)\b(SET)\b(.*)/) &&
             (origContentUp.match(/\b(FRO(M|\b))\b/) ||
              origContentUp.match(/\b(TO)\b/)) ) ) {
            // redo the whole document whenever there's SET FROM/TO involved
            this.setCheck = {
                from: {
                    FILE: false,
                    location: {
                        ENVIRONMENT: false,
                        SYSTEM: false,
                        SUBSYSTEM: false,
                        TYPE: false,
                        STAGE: false,
                    }
                },
                to: {
                    FILE: false,
                    location: {
                        ENVIRONMENT: false,
                        SYSTEM: false,
                        SUBSYSTEM: false,
                        TYPE: false,
                        STAGE: false,
                    }
                },
            };
            const newContent = origContent.substring(0, start)
                                + newText
                                + origContent.substring(end, origContent.length);
            this.statements = this.parseTextIntoSCLstatementsTokens(newContent);
            this.walkStatements();
            return;
        }

        // otherwise incremental update
        const oldLength = end - start;
        const indexPlus = newText.length - oldLength;

        const updatePositionInStatement = ((indexPlus: number, statement: SCLstatement) => {
            statement.starti = statement.starti + indexPlus;
            statement.endi = statement.endi + indexPlus;
            for (const token of statement.tokens) {
                token.starti = token.starti + indexPlus;
            }
            for (const diag of statement.diagnostics) {
                diag.starti = diag.starti + indexPlus;
                diag.endi = diag.endi + indexPlus;
                // the diag.diagnostic.range will be refreshed after all the updates are processed, and textDocument is updated
            }
        });

        let affectedRange = {
            start,
            end
        }; // the start and end index in the textDocument, of where the change affected
        for (let i = this.statements.length - 1; i >= 0 ; -- i) {
            if (this.statements[i].starti > affectedRange.end) {
                // start end statement.starti statement.endi
                updatePositionInStatement(indexPlus, this.statements[i]);
                continue;
            } else if (this.statements[i].endi < affectedRange.start) {
                // statement.starti statement.endi start end, and statement is a complete statement
                const lastToken = this.statements[i].tokens[this.statements[i].tokens.length-1];
                if (lastToken.is_eoStatement) {
                    continue;
                }
            }
            affectedRange.start = Math.min(
                affectedRange.start, this.statements[i].starti);
            affectedRange.end = Math.max(
                affectedRange.end, this.statements[i].endi);

            // before remove this statement, update SCLDocumentManager.numberOfProblems
            SCLDocumentManager.numberOfProblems = SCLDocumentManager.numberOfProblems - this.statements[i].diagnostics.length;
            if (SCLDocumentManager.numberOfProblems < 0) {
                SCLDocumentManager.numberOfProblems = 0;
            }
            this.statements.splice(i, 1);
        }

        const newTextTobeParse: string = origContent.substring(affectedRange.start, start)
                                            + newText
                                            + origContent.substring(end, affectedRange.end);

        const newStatements: SCLstatement[] = this.parseTextIntoSCLstatementsTokens(newTextTobeParse, affectedRange.start);
        newStatements.forEach((newscl) => {
            this.walkStatement(newscl);
            this.statements.push(newscl);
        });
    }

    /**
     * Push diagnostic information into the corresponded SCLstatement
     *
     * @param {ITokenizedString} diagnosedToken the token where the error/warning happened
     * @param {SCLstatement} statement the scl statement where the error/warning happened
     * @param {DiagnosticSeverity} severity
     * @param {string} message
     * @param {string} [relatedMsg]
     * @returns
     * @memberof SCLDocument
     */
    pushDiagnostic(
        diagnosedToken: ITokenizedString,
        statement: SCLstatement,
        severity: DiagnosticSeverity,
        message: string,
        relatedMsg?: string,
        relatedToken?: ITokenizedString
        ) {

        if (SCLDocumentManager.numberOfProblems >= SCLDocumentManager.config.maxNumberOfProblems) {
            return;
        }
        if (diagnosedToken.starti < statement.starti ||
            diagnosedToken.starti + diagnosedToken.value.length > statement.endi) {
            // The token is not in the scl!!
            return;
        }

        SCLDocumentManager.numberOfProblems ++;

        let diagnostic: Diagnostic = {
            severity: severity,
            range: {
                start: this.textDocument.positionAt(diagnosedToken.starti),
                end: this.textDocument.positionAt(diagnosedToken.starti + diagnosedToken.value.length)
            },
            message: message,
            source: 'Endevor SCL extension'
        };
        if (SCLDocumentManager.capabilities.hasDiagnosticRelatedInformationCapability &&
            !isNullOrUndefined(relatedMsg) && !isNullOrUndefined(relatedToken)) {
            diagnostic.relatedInformation = [
                {
                    location: {
                        uri: this.textDocument.uri,
                        range: {
                            start: this.textDocument.positionAt(relatedToken.starti),
                            end: this.textDocument.positionAt(relatedToken.starti + relatedToken.value.length),
                        }
                    },
                    message: relatedMsg,
                }
            ];
        }
        statement.diagnostics.push({
            diagnostic,
            starti: diagnosedToken.starti,
            endi: diagnosedToken.starti + diagnosedToken.value.length,
        });
    }

    private formatStatement(statement: SCLstatement): TextEdit[] {
        const tokens: ITokenizedString[] = statement.tokens;
        const edits: TextEdit[] = [];

        const checkIfKeyword = ((token: ITokenizedString) => {
            Object.entries(ParserTags).forEach(([key, value]) => {
                if (matchWithoutDiagnose(token, value)) {
                    token.is_keyword = true;
                    return;
                }
            });
        }); // do not rely on parser because scl can be invalid, and parser will quit before it goes to the end

        // Step 1, make line endings after value, make only 1 space between each token
        for (let i = 0; i < tokens.length-1; ++ i) {
            checkIfKeyword(tokens[i]);
            if (tokens[i].value === "," || tokens[i].value === ")" || tokens[i].value.endsWith(")")) {
                edits.push({
                    range: {
                        start: this.textDocument.positionAt(tokens[i].starti + tokens[i].value.length),
                        end: this.textDocument.positionAt(tokens[i+1].starti),
                    },
                    newText: "\n"
                });
                continue;
            }
            if (!tokens[i].is_keyword) {
                if (tokens[i+1].value !== "," &&
                    tokens[i+1].value !== ")" &&
                    !matchWithoutDiagnose(tokens[i+1], ParserTags.LEVEL)) {
                    edits.push({
                        range: {
                            start: this.textDocument.positionAt(tokens[i].starti + tokens[i].value.length),
                            end: this.textDocument.positionAt(tokens[i+1].starti),
                        },
                        newText: "\n"
                    });
                    continue;
                }
            }
            let starti = tokens[i].starti + tokens[i].value.length;
            let endi = tokens[i+1].starti;
            if (endi < starti)
                starti = endi;
            edits.push({
                range: {
                    start: this.textDocument.positionAt(starti),
                    end: this.textDocument.positionAt(endi),
                },
                newText: " "
            });
        }

        // Step2. indent at special keywords
        const indentKey = [ParserTags.FROM, ParserTags.TO, ParserTags.OPTION,
            ParserTags.THROUGH, ParserTags.WHERE, ParserTags.VERSION, ParserTags.DATA];
        const searchIndentKey = ((token: ITokenizedString): string => {
            for (const key of indentKey) {
                if (matchWithoutDiagnose(token, key)) {
                    return key;
                }
            }
            return "";
        });
        const indentMap: Map<string, number> = new Map();
        indentMap.set(ParserTags.FROM, 9);
        indentMap.set(ParserTags.TO, 7);
        indentMap.set(ParserTags.OPTION, 12);
        let indent = 0;
        for (let i = 1; i < tokens.length-1; ++ i) {
            if (tokens[i].is_keyword) {
                const searchKey = searchIndentKey(tokens[i]);
                if (searchKey.length > 0) {
                    edits.push({
                        range: {
                            start: this.textDocument.positionAt(tokens[i].starti),
                            end: this.textDocument.positionAt(tokens[i].starti),
                        },
                        newText: " ".repeat(4)
                    });
                    const newIndent = indentMap.get(searchKey);
                    if (!isNullOrUndefined(newIndent)) {
                        indent = newIndent;
                    }
                    continue;
                }
                if (!tokens[i-1].is_keyword && !tokens[i].value.match(/[,()]/)) {
                    edits.push({
                        range: {
                            start: this.textDocument.positionAt(tokens[i].starti),
                            end: this.textDocument.positionAt(tokens[i].starti),
                        },
                        newText: " ".repeat(indent)
                    });
                }
            } else {
                if (tokens[i-1].value === ",")
                    edits.push({
                        range: {
                            start: this.textDocument.positionAt(tokens[i].starti),
                            end: this.textDocument.positionAt(tokens[i].starti),
                        },
                        newText: " ".repeat(indent)
                    });
            }
        }
        return edits;
    }

    formatDocument(): TextEdit[] {
        let edits: TextEdit[] = [];
        edits.push({
            range: {
                start: this.textDocument.positionAt(0),
                end: this.textDocument.positionAt(this.statements[0].starti),
            },
            newText: ""
        });

        let i = 0;
        for (i = 0; i < this.statements.length-1; ++ i) {
            edits = edits.concat(this.formatStatement(this.statements[i]));
            edits.push({
                range: {
                    start: this.textDocument.positionAt(this.statements[i].endi),
                    end: this.textDocument.positionAt(this.statements[i+1].starti),
                },
                newText: "\n\n"
            });
        }
        edits = edits.concat(this.formatStatement(this.statements[i]));
        edits.push({
            range: {
                start: this.textDocument.positionAt(this.statements[i].endi),
                end: this.textDocument.positionAt(this.statements[i].endi),
            },
            newText: "\n"
        });
        return edits;
    }
}