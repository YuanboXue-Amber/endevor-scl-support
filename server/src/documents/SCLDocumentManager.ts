import {
    TextDocument,
    TextDocumentItem,
    VersionedTextDocumentIdentifier,
    TextDocumentContentChangeEvent,
    TextDocumentPositionParams,
    CompletionItem,
    TextDocumentIdentifier,
    TextEdit,
    CompletionItemKind,
    Position, Range,
    Command,
    CodeLens,
    ExecuteCommandParams} from "vscode-languageserver";
import { SCLDocument } from './SCLDocument';
import { isNull, isNullOrUndefined } from "util";
import { isUnaryLike } from "@babel/types";
import { commands, executeSubmitSCL } from "../ExecuteCommandProvider";

export interface IDocumentSettings {
    maxNumberOfProblems: number;
    isREST: boolean;
}

interface ICapabilities {
    hasDiagnosticRelatedInformationCapability: boolean;
}

export const actionCompletion = [ "SET", "ADD", "UPDATE", "DELETE", "GENERATE",
"MOVE", "RETRIEVE", "SIGNIN", "TRANSFER", "APPROVE", "DENY", "BACKIN",
"BACKOUT", "CAST", "DEFINE", "EXECUTE", "RESET", "COMMIT", "LIST"];

/**
 * Manage all the opened documents in the client
 *
 * @export
 * @class SCLDocumentManager
 */
export class SCLDocumentManager {
    static config: IDocumentSettings = { maxNumberOfProblems: 1000, isREST: false };
    static capabilities: ICapabilities = { hasDiagnosticRelatedInformationCapability: false };

    static numberOfProblems = 0;

    // a map of document uri to document details
    documents: Map<string, SCLDocument> = new Map();

    constructor() {
        this.documents = new Map();
    }

    // NEVER USED
    openDocument(textDocumentItem: TextDocumentItem): SCLDocument {
        const textDocument: TextDocument = TextDocument.create(
            textDocumentItem.uri, textDocumentItem.languageId, textDocumentItem.version, textDocumentItem.text);
        const document: SCLDocument = new SCLDocument(textDocument);
        this.documents.set(textDocument.uri, document);

        // no need to refresh the diagnose number, since it is dealt with in SCLDocument.pushDiagnostic
        return document;
    }

    openOrChangeDocument(textDocument: TextDocument): SCLDocument {
        let document = this.documents.get(textDocument.uri);
        if (!document) {
            document = new SCLDocument(textDocument);
            this.documents.set(textDocument.uri, document);
            return document;
        }

        // reset the diagnose number
        let decreaseNum = 0;
        document.statements.forEach((statement) => {
            decreaseNum += statement.diagnostics.length;
        });
        SCLDocumentManager.numberOfProblems = SCLDocumentManager.numberOfProblems - decreaseNum;
        if (SCLDocumentManager.numberOfProblems < 0) {
            SCLDocumentManager.numberOfProblems = 0;
        }

        document.textDocument = textDocument;
        document.fullUpdate();
        this.documents.set(textDocument.uri, document);

        return document;
    }

    closeDocument(textDocumentUri: string): void {
        const document = this.documents.get(textDocumentUri);
        if (isNullOrUndefined(document)) {
            return;
        }

        // refresh the diagnose number
        let decreaseNum = 0;
        document.statements.forEach((statement) => {
            decreaseNum += statement.diagnostics.length;
        });
        SCLDocumentManager.numberOfProblems = SCLDocumentManager.numberOfProblems - decreaseNum;
        if (SCLDocumentManager.numberOfProblems < 0) {
            SCLDocumentManager.numberOfProblems = 0;
        }

        this.documents.delete(textDocumentUri);
    }

    getCompletionBySyntax(textDocumentPosition: TextDocumentPositionParams): CompletionItem[] {
        const document = this.documents.get(textDocumentPosition.textDocument.uri);
        if (!document) {
            throw new Error(`Cannot provide completion on unopened document: ${textDocumentPosition.textDocument.uri}`);
        }

        const tobeCompelteIndex = document.textDocument.offsetAt(textDocumentPosition.position);
        let completionItems: CompletionItem[] = [];

        for (const statement of document.statements) {
            if (tobeCompelteIndex >= statement.starti && tobeCompelteIndex <= statement.endi) {
                let i = -1;
                while (i < statement.tokens.length - 1) {
                    i ++;
                    const starti = statement.tokens[i].starti + statement.tokens[i].value.length;
                    const endi = (i+1) < statement.tokens.length ? statement.tokens[i+1].starti : statement.endi;

                    if (tobeCompelteIndex >= starti && tobeCompelteIndex <= endi) {
                        // to complete between tokens[i] and tokens[i+1] in scl
                        const values = statement.tokens[i].completionItems;
                        if (!isNullOrUndefined(values)) {
                            if (statement.tokens[i].value !== ".") {
                                completionItems = completionItems.concat(values);
                            } else {
                                for (const each of actionCompletion) {
                                    completionItems.push({
                                        label: each + " ",
                                        kind: CompletionItemKind.Function,
                                        documentation: "Endevor SCL keyword"
                                    });
                                }
                            }
                        }
                        return completionItems;
                    }
                }
            }
        }

        // to complete at the end of scl
        const statement = document.statements[document.statements.length-1];
        let values;
        if (!isNullOrUndefined(statement))
            values = statement.tokens[statement.tokens.length-1].completionItems;
        if (!isNullOrUndefined(values) && statement.tokens[statement.tokens.length-1].value !== ".") {
            completionItems = completionItems.concat(values);
        } else {
            // maybe a new scl is started
            for (const each of actionCompletion) {
                completionItems.push({
                    label: each + " ",
                    kind: CompletionItemKind.Function,
                    documentation: "Endevor SCL keyword"
                });
            }
        }
        return completionItems;
    }

    async formatDocument(textDocument: TextDocumentIdentifier): Promise<TextEdit[]> {
        const document = this.documents.get(textDocument.uri);
        if (!document) {
            throw new Error(`Cannot provide formatting on unopened document: ${textDocument.uri}`);
        }

        return document.formatDocument();
    }

    async computeCodeLenses(textDocument: TextDocumentIdentifier): Promise<CodeLens[]> {
        const document = this.documents.get(textDocument.uri);
        if (!document) {
            throw new Error(`Cannot provide codeLens on unopened document: ${textDocument.uri}`);
        }

        let codeLens: CodeLens[] = [];
        for (const statement of document.statements) {
            const range = {
                start: document.textDocument.positionAt(statement.starti),
                end: document.textDocument.positionAt(statement.starti+1),
            };
            codeLens = codeLens.concat(
                commands.map(command => ({
                    range,
                    command: Command.create(command.title, command.command, textDocument, range)
                }))
            );
        }
        return codeLens;
    }

    async executeCodeLens(params: ExecuteCommandParams) {
        const textDocument: TextDocumentIdentifier = (params.arguments as any[])[0];
        const range: Range = (params.arguments as any[])[1];
        const document = this.documents.get(textDocument.uri);
        if (!document) {
            throw new Error(`Cannot provide codeLens on unopened document: ${textDocument.uri}`);
        }
        return executeSubmitSCL(document, document.textDocument.offsetAt(range.start));
    }

}