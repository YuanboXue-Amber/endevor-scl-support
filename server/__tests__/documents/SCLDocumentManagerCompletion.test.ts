import { SCLDocumentManager } from '../../src/documents/SCLDocumentManager';
import { TextDocumentItem, TextDocumentContentChangeEvent, CompletionItem, TextDocumentPositionParams } from 'vscode-languageserver';
import { SCLDocument } from '../../src/documents/SCLDocument';

describe("Test", () => {

    it("test1: getCompletionBySyntax", async () => {

        const openedDocument: TextDocumentItem = {
            uri: "mocked/uri",
            languageId: "mocklanguageId",
            version: 1,
            text: "SET action add . \n SET ACTION RESTORE. \n Set INvalid. "
        };
        const manager = new SCLDocumentManager();
        const document: SCLDocument = manager.openDocument(openedDocument);
        expect(document.textDocument.getText()).toMatch(openedDocument.text);

        const textDocumentPosition: TextDocumentPositionParams = {
            textDocument: {
                uri: openedDocument.uri
            },
            position: {
                line: 2,
                character: 5
            }
        };
        const completions: CompletionItem[] = manager.getCompletionBySyntax(textDocumentPosition);
        expect(completions).toMatchSnapshot();
    });

    it("test2: getCompletionBySyntax", async () => {

        const openedDocument: TextDocumentItem = {
            uri: "mocked/uri",
            languageId: "mocklanguageId",
            version: 1,
            text: "SET action  . \n SET ACTION RESTORE. \n Set  "
        };
        const manager = new SCLDocumentManager();
        const document: SCLDocument = manager.openDocument(openedDocument);
        expect(document.textDocument.getText()).toMatch(openedDocument.text);

        const textDocumentPosition: TextDocumentPositionParams = {
            textDocument: {
                uri: openedDocument.uri
            },
            position: {
                line: 2,
                character: 6
            }
        };
        const completions: CompletionItem[] = manager.getCompletionBySyntax(textDocumentPosition);
        expect(completions).toMatchSnapshot();

        const textDocumentPosition2: TextDocumentPositionParams = {
            textDocument: {
                uri: openedDocument.uri
            },
            position: {
                line: 0,
                character: 11
            }
        };
        const completions2: CompletionItem[] = manager.getCompletionBySyntax(textDocumentPosition2);
        expect(completions2).toMatchSnapshot();
    });

});