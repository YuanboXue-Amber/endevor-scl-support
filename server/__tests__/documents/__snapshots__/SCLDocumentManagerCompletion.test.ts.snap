// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test test1: getCompletionBySyntax 1`] = `
Array [
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "ACTION",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "BUILD",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "FROM",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "OPTIONS",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "STOPRC",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "TO",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "WHERE",
  },
]
`;

exports[`Test test2: getCompletionBySyntax 1`] = `
Array [
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "ACTION",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "BUILD",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "FROM",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "OPTIONS",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "STOPRC",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "TO",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "WHERE",
  },
]
`;

exports[`Test test2: getCompletionBySyntax 2`] = `
Array [
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "ADD",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "ARCHIVE",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "COPY",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "DELETE",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "GENERATE",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "LIST",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "MOVE",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "PRINT",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "RESTORE",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "RETRIEVE",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "SIGNIN",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "TRANSFER",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "UPDATE",
  },
  Object {
    "documentation": "Endevor SCL keyword",
    "kind": 1,
    "label": "VALIDATE",
  },
]
`;
